### ISSUE TITLE
`$_SERVER['SERVER_NAME']` should be not used

### ISSUE THREAD

**Issue by mvorisek** [PrestaShop/PrestaShop#24624]
#### Describe the bug

`$_SERVER['SERVER_NAME']` is not reliable and contains very often a different name than `$_SERVER['HTTP_HOST']`, sometimes port is present, sometimes it is the full domain name, sometimes only the configured/static server name like `localhost` or `127.0.0.1`.

Luckily, the usage of it is minimal, also of `Tools::getServerName()` which uses it.

#### Expected behavior

No use of `$_SERVER['SERVER_NAME']` and `$_SERVER['HTTP_HOST']`.

**Comment by hibatallahAouadni** [comment]
Hello @mvorisek 

Thanks for your issue. Unfortunately, there is not enough information provided to reproduce it and work on it. **You must follow the template to submit a complete bug report**.

Read more about how we expect the issues to be handled [here](https://build.prestashop.com/news/how-bug-reports-are-handled/). Feel free to create a new issue if you can provide further information, we will be glad to investigate it.

Thanks!

**Comment by mvorisek** [comment]
What do you want here more than a report that `$_SERVER['SERVER_NAME']` data should be not used because of the reasons described?

**Comment by matks** [comment]
Hello @mvorisek I think what @hibatallahAouadni means is that this is not a bug report.

A bug report is "component A of PrestaShop should behave like ... but instead it behaves like ... ; so it is faulty, this is not what is expected".

What you are reporting is, I think, an improvement suggestion for the codebase, which is not to rely on $_SERVER['SERVER_NAME'] and $_SERVER['HTTP_HOST'] . We have a issue template for improvements https://github.com/PrestaShop/PrestaShop/issues/new?assignees=&labels=&template=2_feature_request.md

**Comment by mvorisek** [comment]
It is a bug as some code does not work properly if the SERVER_NAME does not match HTTP_HOST.

Partial fix is here https://github.com/PrestaShop/PrestaShop/pull/24626

**Comment by hibatallahAouadni** [comment]
Kind of related to #22581

**Comment by mvorisek** [comment]
Yes - only PrestaShop shop domain must be used to build URLs.

I would vote to deprecate `Tools::getServerName()` method as the data source is unreliable (can be for ex. `localhost` or `my_server`). But PRs it this sense is often rejected, so guy, decide on your own.

Also, I noticed some code stores domain/path settings during installation and when the config is updated after the install, not all paths are correct. Should be tested in CI (install using completely different domain/URL, change config in DB and run all behat tests).

**Comment by hibatallahAouadni** [comment]
@mvorisek could you please attach/link your PR to this issue #22581 and please add a comment in the issue (explaining all your concerns about this problem).
Thanks in advance!

**Comment by matks** [comment]
This issue is important in my opinion so I will provide a reproductible bug report following [the guidelines written here](https://github.com/PrestaShop/PrestaShop/pull/23979#issuecomment-906376989).

**Comment by matks** [comment]
Interesting finding so far: [SERVER_NAME Versus HTTP_HOST
](https://shiflett.org/blog/2006/server-name-versus-http-host)

If the server runs on a port other than 80 then HTTP_HOST contains the port, while SERVER_NAME does not.

**Comment by matks** [comment]
So, after exploration of this matter, I did not find a clear issue using `$_SERVER['SERVER_NAME']` in `Tools::getServerName()` at least in regular setups.

`Tools::getServerName()` is used in PrestaShop in 2 locations:
1. in `Tools::urlBelongsToShop()`
2. in `AdminLoginControllerCore::initContent()`

The main usage is location 1.

[In `Tools::urlBelongsToShop()`](https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Tools.php#L487) ; `Tools::getServerName()` is used to compare the `host` of a given URL to servername. This host is fetched using [parse_url](https://www.php.net/manual/en/function.parse-url.php) which purposefully extracts the port from the hostname.

So we could say that, although `Tools::getServerName()` could have a better name, it returns the right value if we look at its usage, which is the host without the port. It's fine to compare the 2 values if we are sure both omit the port and this is the expected behavior. For most webserver setups, this seems to be just fine.

> It is a bug as some code does not work properly if the SERVER_NAME does not match HTTP_HOST.

Is there an example of such logic in PrestaShop where the difference in behavior between SERVER_NAME and HTTP_HOST does trigger un unwanted behavior?

I have explored some setups, including running the application on a port other than 80 and did not observe issues.

> I would vote to deprecate Tools::getServerName() method as the data source is unreliable (can be for ex. localhost or my_server).

It depends on what you expect from `Tools::getServerName()`. This method actually does exactly what it says: return the configured server name. What could be improved is its usage.

Without an accurate scenario and setup that demonstrates the issue of using  `$_SERVER['SERVER_NAME']` we are left with a theoretical issue.

I am open to continue exploring this topic if given some more informations on an accurate scenario and setup where the usage of `$_SERVER['SERVER_NAME']` triggers a bug in the application usage.
